plugins/cpp_reparse/service/src/astcache.h

```diff
@@ -0,0 +1,118 @@
+#ifndef CC_SERVICE_CPPREPARSESERVICE_ASTCACHE_H
+#define CC_SERVICE_CPPREPARSESERVICE_ASTCACHE_H
+
+#include <chrono>
+#include <map>
+#include <memory>
+#include <mutex>
+
+// Required for the Thrift objects, such as core::FileId.
+#include "cppreparse_types.h"
+
+namespace clang
+{
+class ASTUnit;
+} // namespace clang
+
+namespace cc
+{
+
+namespace service
+{
+
+namespace reparse
+{
+
+/**
+ * Implements a caching mechanism over Clang ASTs to ease the runtime overhead
+ * of having to parse source files over and over again, as it is a *very*
+ * costly operation.
+ *
+ * This class owns the ASTUnit instances cached. It is expected that this
+ * class outlives the execution of FrontendActions over an AST.
+ */
+class ASTCache
+{
+public:
+
+  /**
+   * @param maxCacheSize_ The maximum size of the cache above which automatic
+   * pruning of old entries will take place. This is not an absolute limit, the
+   * cache is allowed to overfill in case no more entries could be pruned.
+   */
+  ASTCache(size_t maxCacheSize_);
+
+  ASTCache(const ASTCache&) = delete;
+  ASTCache& operator=(const ASTCache&) = delete;
+  ~ASTCache() = default;
+
+  /**
+   * Retrieves the AST stored for the given file ID, or a nullptr if none is
+   * stored.
+   */
+  std::shared_ptr<clang::ASTUnit> getAST(const core::FileId& id_);
+
+  /**
+   * Store the AST for the given file in the cache. The AST Cache takes
+   * ownership over the ASTUnit. The method returns the shared pointer that
+   * wraps the ASTUnit argument given.
+   */
+  std::shared_ptr<clang::ASTUnit> storeAST(
+    const core::FileId& id_,
+    std::unique_ptr<clang::ASTUnit>&& AST_);
```

Comment:
Why are `std::unique_ptr`s passed by `&&`?

Reply:
`std::unique_ptr` cannot be copy constructed, thus cannot be passed by value.

Clang gives you (the client code) unique-ptrs of ASTs. The extra requirement for the user to **move the AST *into* the method** is to signify that `storeAST` *destroys* the unique ownership and internally transforms it into a `shared_ptr` (see `ASTCacheEntry`). Thus, methods that create the ASTs explicitly release the unique_ptr they had in their stack by turning the unique-ptr on the stack into a *moved from* state.

Simply passing the `unique_ptr` by reference wouldn't make this rule explicit enough.

(Later code uses the cache's get method to obtain the same AST back, but that obtained AST is through a shared pointer to ensure ownership is kept as long as the cache can store the AST or some thread is using it.)

Reply:
Yes, the idea behind the mechanism is understandable with ownership, shared pointers, etc. But the parameter passing method is not justified. For moving a `std::unique_ptr` you don't need `&&` as formal parameter, since it has a move constructor:
```.cpp
struct S {};

void f(std::unique_ptr<S>) {}

void g()
{
  std::unique_ptr<S> p;
  f(std::move(p));
}
```

Reply:
I'll update the previous patch, as these things were brought in by that patch, not this one. Unfortunately GitHub can't cherry patches incrementally.

Reply:
Has been fixed in #248.